crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+				}
+				Some(Listing::Auction(listing)) => {
+					if let Some((winner, hammer_price)) = ListingWinningBid::<T>::take(&collection_id, token_id) {
+						if let Err(err) =
+							Self::settle_auction(&collection_id, token_id, &listing, &winner, hammer_price)
+						{
+							// auction settlement failed despite our prior validations.
+							// release winning bid tokens. listing metadadta is removed by now.
+							log!(error, "🃏 auction settlement failed: {:?}", err);
+							T::MultiCurrency::unreserve(&winner, Some(listing.payment_asset), hammer_price);
+
+							Self::deposit_event(RawEvent::AuctionClosed(
+								collection_id.clone(),
+								token_id,
+								AuctionClosureReason::SettlementFailed,
+							));
+						} else {
+							// auction settlement success
+							Self::deposit_event(RawEvent::AuctionSold(
+								collection_id.clone(),
+								token_id,
+								listing.payment_asset,
+								hammer_price,
+								winner,
+							));
+						}
+					} else {
+						// normal closure, no acceptable bids
+						Self::deposit_event(RawEvent::AuctionClosed(
+							collection_id.clone(),
+							token_id,
+							AuctionClosureReason::ExpiredNoBids,
+						));
+					}
+				}
+				_ => (),
+			}
+			removed += 1;
+		}
+
+		removed
+	}
+	/// Settle an auction listing (guaranteed to be atomic).
+	/// - transfer funds from winning bidder to entitled royalty accounts and seller
+	/// - transfer ownership to the winning bidder
+	#[transactional]
+	fn settle_auction(
+		collection_id: &[u8],
+		token_id: T::TokenId,
+		listing: &AuctionListing<T>,
+		winner: &T::AccountId,
+		hammer_price: Balance,
+	) -> DispatchResult {
+		// if there are no custom royalties, fallback to default if it exists
+		let royalties_schedule = if let Some(royalties_schedule) = Self::token_royalties(collection_id, token_id) {
+			royalties_schedule
+		} else {
+			Self::collection_royalties(collection_id).unwrap_or_else(Default::default)
+		};
+
+		let for_royalties = royalties_schedule.calculate_total_entitlement() * hammer_price;
+		let mut for_seller = hammer_price;
+
+		// do royalty payments
+		if !for_royalties.is_zero() {
+			for (who, entitlement) in royalties_schedule.entitlements.into_iter() {
+				let royalty = entitlement * hammer_price;
+				let _ = T::MultiCurrency::repatriate_reserved(&winner, Some(listing.payment_asset), &who, royalty)?;
+				for_seller -= royalty;
+			}
+		}
+
+		let current_owner = Self::token_owner(collection_id, token_id);
+		let seller_balance = T::MultiCurrency::free_balance(&current_owner, Some(listing.payment_asset));
+		let _ =
+			T::MultiCurrency::repatriate_reserved(&winner, Some(listing.payment_asset), &current_owner, for_seller)?;
+
+		// The implementation of `repatriate_reserved` may take less than the required amount and succeed
+		// this should not happen but could for reasons outside the control of this module
+		ensure!(
```

Comment:
A late error, after royalties are handled, is detrimental to the state of the chain. Please take an approach that wouldn't fail here at all and move any checks to the beginning of this function.

Reply:
please read about `#[transactional]`

Reply:
Is there any overhead with using #[transactional]?

Reply:
yes

Reply:
So let's use it when avoiding an incomplete state is very hard or due to using other libraries out of our control. We seem to be able to re-arrange our implementation in this case and some other ones and get to the same guarantees with no extra cost. An example here is you check how much is actually reserved from the seller at the beginning of this functopn before proceeding with repatriation. Anyway, I can understand if you want to leave it to another round of refinement/improvement.